peer/abstractlist/list_test.go

```diff
@@ -209,21 +219,68 @@ func TestPeerList(t *testing.T) {
 	// Invalid updates, after stop
 	assert.Error(t, list.Update(peer.ListUpdates{
 		Additions: []peer.Identifier{
-			hostport.Identify("3.3.3.3:4040"),
+			abstractpeer.Identify("3.3.3.3:4040"),
 		},
 		Removals: []peer.Identifier{
-			hostport.Identify("4.4.4.4:4040"),
+			abstractpeer.Identify("4.4.4.4:4040"),
 		},
 	}))
 
 	assert.NoError(t, list.Update(peer.ListUpdates{
 		Additions: []peer.Identifier{},
 		Removals: []peer.Identifier{
-			hostport.Identify("3.3.3.3:4040"),
+			abstractpeer.Identify("3.3.3.3:4040"),
 		},
 	}))
 }
 
+func TestFailWait(t *testing.T) {
+	fake := yarpctest.NewFakeTransport(yarpctest.InitialConnectionStatus(peer.Available))
+	impl := &mraList{}
+	list := New("mra", fake, impl)
+
+	require.NoError(t, list.Start())
+
+	// This case induces the list to enter the wait loop until a peer is added.
+
+	go func() {
+		time.Sleep(10 * testtime.Millisecond)
+		if err := list.Update(peer.ListUpdates{
+			Additions: []peer.Identifier{
+				abstractpeer.Identify("0"),
+			},
+		}); err != nil {
+			t.Log(err.Error())
+			t.Fail()
+		}
+	}()
+
+	{
+		ctx, cancel := context.WithTimeout(context.Background(), testtime.Second)
+		defer cancel()
+
+		p, onFinish, err := list.Choose(ctx, &transport.Request{})
+		require.NoError(t, err)
+		onFinish(nil)
+
+		assert.Equal(t, "0", p.Identifier())
```

Comment:
Can we use a `require` here or maybe better, change the scopes to subtests?

Reply:
I think I would like to keep it. Failing wouldn’t lead to a panic later in the test.